### PR TITLE
Build Simulations from Config

### PR DESCRIPTION
**Pull request by dog-ball**

### CODE REVIEW

**Review by j-ti — COMMENTED** [review]
- it is good to regard it as an own module e.g. "config" in this case (potentially to be transfered to a distinct repo)
- avoid uploading data to github (if needed only a minimal example files)
- data generally has to be placed manually or alternatively paths to data need to be set in the config file => share 
- network json/png should not be in simply module folder and if it is an output it should not be in the repo

**Review by j-ti — CHANGES_REQUESTED** [review]
- always use of pathlib and parameters from config files
- normalize time series and correct scaling parameter
- split simply functions if reusability is achieved
- dataframe indexing

File: config/build_simulation.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    config_df = read_config_json(config_json)
+    # Create nodes for power network
+    pn = create_power_network_from_config(network_path, weight_factor)
+
+    if plot_network is True:
+        pn.plot()
+
+    actors = []
+    file_dict = {}
+    asset_dict = {}
```
Comment (j-ti) [inline config/build_simulation.py:108]: dictionaries have to be reseted for each actor

Comment (j-ti) [reply config/build_simulation.py:108]: ... accordingly the asset data has to be set to zero or the actor has to interpret the nan values as such

Comment (dog-ball) [reply config/build_simulation.py:108]: Don't know how I missed this! Thanks

File: config/build_simulation.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+    # Build config object using configuration file
+    cfg = Config(args.config)
+    cfg.path = Path('../') / cfg.path
+    # Reset save location directory
+    remove_existing_dir(cfg.path)
+    sc = create_scenario_from_config(args.scenario_config, args.network, data_dirpath=args.data_dir,
+                                     nb_ts=cfg.nb_ts, loads_dir_path=args.loads_dir, ps=1, ls=None)
+    sc.save(cfg.path, cfg.data_format)
+    update_actor_json_for_market_maker(cfg.path)
```
Comment (j-ti) [inline config/build_simulation.py:217]: is this function call needed when the config prosumerName is already "market_maker"? see comment at function

File: main.py [hunk 3]
```diff
@@ -51,12 +52,12 @@
 
     # generate requested market
     if "pac" in cfg.market_type:
-        m = market_2pac.TwoSidedPayAsClear(0)
+        m = market_2pac.TwoSidedPayAsClear(0, network=sc.power_network)
```
Comment (j-ti) [inline main.py:55]: Includes clusters to orders, but without influence on the prices as this is done during matching.
- This should be tackled within another PR

File: config/build_simulation.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+    if cfg.show_plots:
+        sc.power_network.plot()
+        sc.plot_actor_data()
+
+    if cfg.save_network:
```
Comment (j-ti) [inline config/build_simulation.py:223]: should both be part of sc.save(), adapt function by:
- adding power_network.to_image()
- adapting code to use power_network.to_json() and implicit changes to the load function in scenario.py

File: simply/power_network.py [hunk 5]
```diff
@@ -215,3 +215,15 @@ def load_network():
...
+def create_power_network_from_config(network_path, weight_factor=1):
+    with open(network_path) as user_file:
+        file_contents = user_file.read()
+    network_json = json.loads(file_contents)
+    network_name = list(network_json.keys())[0]
+    network_json = list(network_json.values())[0]
+    network = json_graph.node_link_graph(network_json,
+                                         directed=network_json.get("directed", False),
+                                         multigraph=network_json.get("multigraph", False))
+    return PowerNetwork(network_name, network, weight_factor)
```
Comment (j-ti) [inline simply/power_network.py:229]: function should be used in Scenario load function, 
while harmonizing with power_network.to_json()

File: simply/market.py [hunk 6]
```diff
@@ -32,8 +32,10 @@ def __init__(self, time, network=None, grid_fee_matrix=None):
             self.grid_fee_matrix = network.grid_fee_matrix
         self.EPS = 1e-10
         if self.save_csv:
-            match_header = ("time", "bid_id", "ask_id", "bid_actor", "ask_actor", "bid_cluster",
-                            "ask_cluster", "energy", "price")
+            match_header = ["time", "bid_id", "ask_id", "bid_actor", "ask_actor", "bid_cluster",
+                            "ask_cluster", "energy", "price"]
+            if len(self.grid_fee_matrix) > 1:
+                match_header.append('grid_fee')
```
Comment (j-ti) [inline simply/market.py:38]: could be added in all cases analog to the cluster ids. Values can then be 0 or None

Comment (j-ti) [reply simply/market.py:38]: maybe rename to included_grid_fee

File: simply/market.py [hunk 7]
```diff
@@ -201,5 +204,17 @@ def create_csv(self, filename, headers):
             writer = csv.writer(f)
             writer.writerow(headers)
 
+    def process_matches_for_csv(self, matches):
+        if self.grid_fee_matrix and len(self.grid_fee_matrix) > 1:
```
Comment (j-ti) [inline simply/market.py:208]: - Docstring missing
- function name too generic, rather: `def add_grid_fee_info()`

File: simply/market.py [hunk 8]
```diff
@@ -201,5 +204,17 @@ def create_csv(self, filename, headers):
...
+            output = []
+            for match in matches:
+                if match['bid_cluster'] is not None and match['ask_cluster'] is not None:
+                    match['grid_fee'] = \
+                        self.grid_fee_matrix[match['bid_cluster']][match['ask_cluster']]
```
Comment (j-ti) [inline simply/market.py:213]: could be a handy function `self.get_grid_fee(match)`

File: simply/market_2pac.py [hunk 9]
```diff
@@ -66,7 +66,8 @@ def match(self, show=False):
         if show:
             print(matches)
 
-        self.append_to_csv(matches, 'matches.csv')
+        output = self.process_matches_for_csv(matches)
```
Comment (j-ti) [inline simply/market_2pac.py:69]: see market.py

File: simply/market_fair.py [hunk 10]
```diff
@@ -241,6 +237,7 @@ def match(self, show=False):
         if show:
             print(matches)
 
-        self.append_to_csv(matches, 'matches.csv')
+        output = self.process_matches_for_csv(matches)
```
Comment (j-ti) [inline simply/market_fair.py:240]: see market.py

File: simply/config.py [hunk 11]
```diff
@@ -84,8 +84,12 @@ def __init__(self, cfg_file):
         self.reset_market = parser.getboolean("default", "reset_market", fallback=True)
         # always create new scenario in given path
         self.update_scenario = parser.getboolean("default", "update_scenario", fallback=False)
+        self.save_network = parser.getboolean("default", "save_network", fallback=False)
```
Comment (j-ti) [inline simply/config.py:87]: see above: not necessary, should be saved by default in sc.save()

File: config/config.txt [hunk 12]
```diff
@@ -0,0 +1,46 @@
+[default]
```
Comment (j-ti) [inline config/config.txt:1]: this is an example file of simply config, and should not be here, if necessary, add relevant information to the config in the example folder of simply

**Review by j-ti — CHANGES_REQUESTED** [review]
Thank you, we are almost done, please see comments

could not finish build_simulation.py yet but already wanted to submit review

File: simply/scenario.py [hunk 13]
```diff
@@ -78,6 +78,9 @@ def save(self, dirpath, data_format):
         # save map_actors
         dirpath.joinpath('map_actors.json').write_text(json.dumps(self.map_actors, indent=2))
 
+        self.power_network.to_image()
+        self.power_network.to_json()
```
Comment (j-ti) [inline simply/scenario.py:82]: - network json is already output at line 56 (-> delete to_json)
- to_image writes to the folder from which the script is called, but should be saved within the  scenario folder

File: config/build_simulation.py [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+    pn.update_shortest_paths()
+    pn.generate_grid_fee_matrix(weight_factor)
+
+    return Scenario(pn, actors, actor_map)
+
+
+if __name__ == "__main__":
+    # Store default configuration files and directories paths
+    dirname = Path(__file__).parent
+    config_json_path = dirname / Path('community_config.json')
```
[the remark this answers is not in the document]
Comment (j-ti) [reply config/build_simulation.py]: [6e27665](https://github.com/BESTenergytrade/simply/pull/138/commits/6e2766510982eb8494d69a7e9f3e7815da1cb208)

File: config/build_simulation.py [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+
+    # Parse configuration files and directories paths as arguments
+    parser = ArgumentParser(description='Entry point for market simulation')
+    parser.add_argument('scenario_config', nargs='?', default=config_json_path,
+                        help='scenario config json file')
+    parser.add_argument('network', nargs='?', default=network_path, help='network json file')
+    parser.add_argument('config', nargs='?', default=config_path, help='configuration file')
+    parser.add_argument('loads_dir', nargs='?', default=loads_dir_path,
+                        help='loads assignment csv file')
+    parser.add_argument('data_dir', nargs='?', default=data_dirpath, help='data directory')
```
[the remark this answers is not in the document]
Comment (j-ti) [reply config/build_simulation.py]: [6e27665](https://github.com/BESTenergytrade/simply/pull/138/commits/6e2766510982eb8494d69a7e9f3e7815da1cb208)

File: simply/scenario.py (hunk 13 above)
Comment (j-ti) [reply simply/scenario.py]: [6e27665](https://github.com/BESTenergytrade/simply/pull/138/commits/6e2766510982eb8494d69a7e9f3e7815da1cb208)

File: config/build_simulation.py [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+        # Save peak value and normalize time series
+        csv_peak[col] = df[col].max()
+
+    df = basic_strategy(df)
+
+    return Actor(actor_id, df)
+
+
+def read_config_json(config_json):
+    config_df = pd.read_json(config_json)
```
[the remark this answers is not in the document]
Comment (j-ti) [reply config/build_simulation.py]: [6e27665](https://github.com/BESTenergytrade/simply/pull/138/commits/6e2766510982eb8494d69a7e9f3e7815da1cb208)

File: config/build_simulation.py [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+
+    # Read csv files for each asset
+    csv_peak = {}
+    for col, csv_dict in asset_dict.items():
+        # if csv_dict is empty
+        if not csv_dict:
+            continue
+        csv_df = pd.read_csv(csv_dict["csv"], sep=',', parse_dates=['Time'], dayfirst=True,
+                             index_col=['Time'])
+        df[col] = csv_df[start_date:end_date]
```
[the remark this answers is not in the document]
Comment (j-ti) [reply config/build_simulation.py]: [6e27665](https://github.com/BESTenergytrade/simply/pull/138/commits/6e2766510982eb8494d69a7e9f3e7815da1cb208)
